File: SmartCredentials_aOS/oneclickbusinessclient/src/main/java/com/example/oneclickbusinessclient/controllers/OneClickBusinessClientController.java
/*
 * Copyright (c) 2019 Telekom Deutschland AG
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.oneclickbusinessclient.controllers;

import android.app.Activity;
import android.content.Context;
import android.util.Log;

import androidx.appcompat.app.AppCompatActivity;

import com.example.oneclickbusinessclient.firebase.FirebaseTokenProvider;
import com.example.oneclickbusinessclient.recommendation.PreferencesHelper;
import com.example.oneclickbusinessclient.recommendation.Recommendation;
import com.example.oneclickbusinessclient.recommendation.RecommendationConstants;
import com.example.oneclickbusinessclient.rest.MakeRecommendationsBody;
import com.example.oneclickbusinessclient.rest.RecommenderApi;
import com.example.oneclickbusinessclient.rest.RetrofitClient;
import com.example.oneclickbusinessclient.ui.BaseLoadingDialogFragment;
import com.example.oneclickbusinessclient.ui.DialogListener;
import com.example.oneclickbusinessclient.ui.OneClickDialogFragment;

import java.text.DecimalFormat;
import java.util.List;
import java.util.Map;

import de.telekom.smartcredentials.core.api.OneClickApi;
import de.telekom.smartcredentials.core.controllers.CoreController;
import de.telekom.smartcredentials.oneclickbusinessclient.R;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Retrofit;

public class OneClickBusinessClientController implements OneClickApi{

    private final CoreController mCoreController;
    private final Context mContext;
    private final FirebaseTokenProvider firebaseTokenProvider;
    private AppCompatActivity mActivity;

    //  private final PublishSubject<String> recommendationListener;
    //  private final PublishSubject<Object> eventListener;

    private final DialogListener dialogListener = new DialogListener() {
        @Override
        public void onContinue() {
            if(mActivity != null){
                BaseLoadingDialogFragment baseLoadingDialogFragment = BaseLoadingDialogFragment.newInstance(mActivity);
                baseLoadingDialogFragment.show(mActivity.getSupportFragmentManager(), BaseLoadingDialogFragment.TAG);
            }
        }
    };

    public OneClickBusinessClientController(CoreController coreController, Context context) {
        mCoreController = coreController;
        mContext = context;
        firebaseTokenProvider = new FirebaseTokenProvider();
//        recommendationListener = PublishSubject.create();
//
//        eventListener = PublishSubject.create();
//        eventListener.subscribe(message -> {
//            if(message instanceof RecommendationReceivedEvent){
//                recommendationListener.onNext("openOneClickDialogFragment");
//                eventListener.onNext(new RecommendationAcceptedEvent());
//            }else if (message instanceof RecommendationAcceptedEvent){
//                recommendationListener.onNext("openLoadingDialogFragment");
//            }else if(message instanceof TokenResponse){
//                recommendationListener.onNext("tokenResponse - open Portal");
//            }else{
//                recommendationListener.onNext("Unknown event");
//            }
//        },
//            throwable -> recommendationListener.onNext("EXCEPTION")
//        );
    }

    @Override
    public void bind(AppCompatActivity activity){
        mActivity = activity;
    }

    @Override
    public void unbind(){
        mActivity = null;
    }

    @Override
    public void makeRecommendation(List<String> productIds, String firebaseServerKey) {
        RetrofitClient retrofitClient = new RetrofitClient();
        String recommenderUrl = retrofitClient.getRecommenderUrl();
        Retrofit retrofit = retrofitClient.createRetrofitClient(recommenderUrl);
        RecommenderApi recommenderApi = retrofit.create(RecommenderApi.class);
        //    FirebaseTokenProvider firebaseTokenProvider = new FirebaseTokenProvider();

       // Observable.just(firebaseTokenProvider.getToken())

        Observable.create(firebaseTokenProvider)
                .flatMap(token -> {
                    MakeRecommendationsBody body = new MakeRecommendationsBody(token,
                            productIds, firebaseServerKey);
                    return recommenderApi.observeMakeRecommendations(body);
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe();
    }

    @Override
    public void pushRecommendationMessage(Map<String, String> data) {
        String recommendationId = data.get(RecommendationConstants.KEY_RECOMMENDATION_ID);
        String clientId = data.get(RecommendationConstants.KEY_CLIENT_ID);
        Log.d(RecommendationConstants.TAG, "Received recommendation notification with ID " +
                recommendationId + "and client ID " + clientId);

        Recommendation recommendation = new Recommendation(recommendationId, clientId);
        PreferencesHelper preferencesHelper = PreferencesHelper.getInstance(mContext);
        preferencesHelper.saveRecommendation(recommendation);

        showDialog();
        //   eventListener.onNext(new RecommendationReceivedEvent());
    }

    @Override
    public void updateFirebaseToken(String token) {
        firebaseTokenProvider.updateToken(token);
    }

    private void showDialog(){
        if(mActivity != null){
            OneClickDialogFragment oneClickDialogFragment = OneClickDialogFragment.newInstance(R.string.one_click_dialog_title,
                    R.string.one_click_after_description, mActivity, dialogListener);
            oneClickDialogFragment.show(mActivity.getSupportFragmentManager(), OneClickDialogFragment.TAG);
        }
    }

}

File: SmartCredentials_aOS/oneclickbusinessclient/src/main/java/de/telekom/smartcredentials/oneclickbusinessclient/controllers/OneClickBusinessClientController.kt
/*
 * Copyright (c) 2019 Telekom Deutschland AG
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package de.telekom.smartcredentials.oneclickbusinessclient.controllers

import androidx.appcompat.app.AppCompatActivity
import androidx.compose.ui.platform.ComposeView
import de.telekom.smartcredentials.core.api.OneClickApi
import de.telekom.smartcredentials.core.blacklisting.SmartCredentialsFeatureSet
import de.telekom.smartcredentials.core.controllers.CoreController
import de.telekom.smartcredentials.core.logger.ApiLoggerResolver
import de.telekom.smartcredentials.core.responses.FeatureNotSupportedThrowable
import de.telekom.smartcredentials.core.responses.RootedThrowable
import de.telekom.smartcredentials.oneclickbusinessclient.BuildConfig
import de.telekom.smartcredentials.oneclickbusinessclient.OneClickClientConfiguration
import de.telekom.smartcredentials.oneclickbusinessclient.recommendation.Recommendation
import de.telekom.smartcredentials.oneclickbusinessclient.recommendation.RecommendationConstants
import de.telekom.smartcredentials.oneclickbusinessclient.repository.StorageRepo
import de.telekom.smartcredentials.oneclickbusinessclient.rest.MakeRecommendationsBody
import de.telekom.smartcredentials.oneclickbusinessclient.rest.RecommenderApi
import de.telekom.smartcredentials.oneclickbusinessclient.rest.RetrofitClient
import de.telekom.smartcredentials.oneclickbusinessclient.ui.PortalOffer
import de.telekom.smartcredentials.oneclickbusinessclient.ui.UIHandler
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import org.json.JSONException

/**
 * Created by larisa-maria.suciu@endava.com at 22/03/2023
 */
class OneClickBusinessClientController(
    private val mCoreController: CoreController,
    private val mStorageRepository: StorageRepo,
    private val config: OneClickClientConfiguration
) : OneClickApi {

    private var uiHandler = UIHandler(this, config)

    override fun bind(activity: AppCompatActivity) {
        uiHandler.bind(activity)
    }

    override fun setComposeView(composeView: ComposeView?) {
        uiHandler.setComposeView(composeView)
    }

    override fun unbind() {
        uiHandler.unbind()
    }

    @Deprecated("Deprecated in Java")
    override fun makeRecommendation(
        productIds: List<String>
    ) {
        ApiLoggerResolver.logMethodAccess(javaClass.simpleName, "makeRecommendation")
        if (mCoreController.isSecurityCompromised) {
            mCoreController.handleSecurityCompromised()
            throw Exception(RootedThrowable())
        }
        if (mCoreController.isDeviceRestricted(SmartCredentialsFeatureSet.ONE_CLICK_CLIENT)) {
            val errorMessage = SmartCredentialsFeatureSet.ONE_CLICK_CLIENT.notSupportedDesc
            throw Exception(FeatureNotSupportedThrowable(errorMessage))
        }

        if (mStorageRepository.isTokenStorageEmpty()) {
            throw RuntimeException("Firebase token is null")
        }

        val retrofit = RetrofitClient().createRetrofitClient(BuildConfig.RECOMMENDER_URL)
        val recommenderApi = retrofit.create(RecommenderApi::class.java)
        try {
            val token = mStorageRepository.firebaseToken
            Observable.just(token)
                .flatMap { firebaseToken: String ->
                    val body = MakeRecommendationsBody(
                        firebaseToken,
                        productIds,
                        config.firebaseServerKey
                    )
                    recommenderApi.observeMakeRecommendations(body)
                }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe()
        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }


    override fun pushRecommendationMessage(
        data: Map<String, String>
    ) {
        ApiLoggerResolver.logMethodAccess(javaClass.simpleName, "pushRecommendationMessage")
        if (mCoreController.isSecurityCompromised) {
            mCoreController.handleSecurityCompromised()
            throw Exception(RootedThrowable())
        }
        if (mCoreController.isDeviceRestricted(SmartCredentialsFeatureSet.ONE_CLICK_CLIENT)) {
            val errorMessage = SmartCredentialsFeatureSet.ONE_CLICK_CLIENT.notSupportedDesc
            throw Exception(FeatureNotSupportedThrowable(errorMessage))
        }

        val recommendationId: String =
            data[RecommendationConstants.KEY_RECOMMENDATION_ID].toString()
        val clientId: String = data[RecommendationConstants.KEY_CLIENT_ID].toString()
        ApiLoggerResolver.logEvent(
            "Received recommendation notification with ID " + recommendationId + "and client ID " + clientId
        )
        val recommendation = Recommendation(recommendationId, clientId)
        mStorageRepository.saveRecommendation(recommendation)
        ApiLoggerResolver.logEvent("pushRecommendationMessage: clientId = $clientId and recommendationId = $recommendationId")
        uiHandler.onRecommendationReceived()
    }

    override fun updateFirebaseToken(
        token: String
    ) {
        ApiLoggerResolver.logMethodAccess(javaClass.simpleName, "updateFirebaseToken")
        if (mCoreController.isSecurityCompromised) {
            mCoreController.handleSecurityCompromised()
            throw Exception(RootedThrowable())
        }
        if (mCoreController.isDeviceRestricted(SmartCredentialsFeatureSet.ONE_CLICK_CLIENT)) {
            val errorMessage = SmartCredentialsFeatureSet.ONE_CLICK_CLIENT.notSupportedDesc
            throw Exception(FeatureNotSupportedThrowable(errorMessage))
        }
        mStorageRepository.saveToken(token)
    }


    internal fun createPortalOffer(
        token: String
    ) {
        ApiLoggerResolver.logMethodAccess(javaClass.simpleName, "createPortalOffer")
        if (mCoreController.isSecurityCompromised) {
            mCoreController.handleSecurityCompromised()
            throw Exception(RootedThrowable())
        }
        if (mCoreController.isDeviceRestricted(SmartCredentialsFeatureSet.ONE_CLICK_CLIENT)) {
            val errorMessage = SmartCredentialsFeatureSet.ONE_CLICK_CLIENT.notSupportedDesc
            throw Exception(FeatureNotSupportedThrowable(errorMessage))
        }

        if (mStorageRepository.isAnyRecommendationAvailable) {
            ApiLoggerResolver.logInfo("Should open portal on file")
            val recommendation = mStorageRepository.getRecommendation
            mStorageRepository.clearRecommendations()
            try {
                val firebaseToken = mStorageRepository.firebaseToken
                val portalOffer = recommendation?.recommendationId?.let { recommendationId ->

                    PortalOffer.PortalOfferBuilder()
                        .setOperatorToken(token)
                        .setFirebaseId(firebaseToken)
                        .setRecommendationId(recommendationId)
                        .setServerKey(config.firebaseServerKey)
                        .build()
                }
                portalOffer?.let { offer -> uiHandler.onOfferAvailable(offer) }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        } else {
            ApiLoggerResolver.logError(javaClass.name, "Unknown event")
        }
    }
}
